thunder/core/module.py

```diff
@@ -215,6 +224,45 @@ def state_dict(self, *, destination=None, prefix="", keep_vars=False):
                 destination[extra_state_key] = self.get_extra_state()
         return destination
 
+    def original_state_dict(
+        self,
+        *,
+        destination: dict[str, Any] | None = None,
+        prefix: str = "",
+        keep_vars: bool = False,
+    ) -> dict[str, Any]:
+        """Returns the state dict of the transformed :class:`ThunderModule` with reverse transform applied.
+
+        For example, :func:`ThunderModule.state_dict` returns a state dict of sharded tensors if
+        a model is :func:`thunder.distributed.fsdp` applied while :func:`ThunderModule.original_state_dict`
+        returns a state dict of unsharded tensors.
+
+        Args:
+            destination: if given, use this mutable mapping as the dict container.
+            prefix: a prefix for the keys.
+            keep_vars: do not detach
+
+        """
+        module_names = {name for name, _ in self._model.named_modules()}
+        state_dict = self.state_dict(destination=destination, prefix=prefix, keep_vars=keep_vars)
+        state_dict_per_submodule = _convert_state_dict_to_per_module(state_dict, module_names)
+
+        _state_dict_keys = list(state_dict.keys())
```

Comment:
I might be missing something, I think it would be simpler if we took/created the destination and then used `transformed_state_dict = self.state_dict(...)` (so not populate this into the destination) and then finally used something like
```
destination.update((f"{prefix}{submodule_prefix}{k}", v) for k, v in submodule_state_dict.items())
```
The background is that Transforms like Quantization, LoRA etc. will change what keys are in the state dict.
I don't think the preserving order in the state dict is important.